proposals/text/6001-document-store-filter-rework.md

```diff
@@ -0,0 +1,197 @@
+
+Comparison dictionaries must contain the keys:
+
+- `field`
+- `operator`
+- `value`
+
+Logic dictionaries have only one mandatory key:
+
+- `conditions`
+
+`conditions` key must be a list of dictionaries, either Comparison or Logic.
+
+The following key can be omitted:
+
+- `operator`
+
+`operator` values in Comparison dictionaries must be:
+
+- `==`
+- `!=`
+- `>`
+- `>=`
+- `<`
+- `<=`
+- `in`
+- `not in`
+
+`operator` values in Logic dictionaries must be:
+
+- `NOT`
+- `OR`
+- `AND`
+
+If `operator` key in Logic dictionaries is omitted we assume it's an `AND` value.
+
+---
+
+As an example this:
+
+```
+{
+    "$and": {
+        "type": {"$eq": "article"},
+        "$or": {"genre": {"$in": ["economy", "politics"]}, "publisher": {"$eq": "nytimes"}},
+        "date": {"$gte": "2015-01-01", "$lt": "2021-01-01"},
+        "rating": {"$gte": 3},
+    }
+}
+```
+
+would convert to:
+
+```
+{
+    "operator": "AND",
+    "conditions": [
+        { "field": "type", "operator": "==", "value": "article" },
+        {
+            "operator": "OR",
+            "conditions": [
+                { "field": "genre", "operator": "in", "value": ["economy", "politics"] },
+                { "field": "publisher", "operator": "==", "value": "nytimes" },
+            ]
+        },
+        { "field": "date", "operator": ">=", "value": "2015-01-01" },
+        { "field": "date", "operator": "<", "value": "2021-01-01" },
+        { "field": "rating", "operator": ">=", "value": 3 },
+    ]
+}
+```
+
+In Python code:
+
+```python
+type == "article" and (
+    genre in ["economy", "politics"] or publisher == "nytimes"
+) and date >= "2015-01-01" and date < "2021-01-01" and rating >= 3
+```
+
+Dates have been kept as strings but ideally in the new implementation they would be converted to `datetime` instances so the Document Store will be able to convert it to whatever format it needs to actually compare them. As different Document Stores might have different ways of storing the same value it's important that they handle the conversion from Python type to stored type.
+
+Another thing that in my opinion should be changed is the that filtering metadata fields must be explicitly specified for filtering. In the example above all `field`s would be prefixed with `metadata.` to get in return the expected `Document`s. e.g. `date` -> `metadata.date`
+This connects to `Document` implementation and is not the focus of this proposal, but it should be taken into account.
+
+# Drawbacks
+
+The only drawback would be that we need to adapt `MemoryDocumentStore` and `ElasticsearchDocumentStore` to use this new filtering language. These are the only two Document Stores currently created for Haystack 2.x.
+
+# Alternatives
+
+An alternative would be keeping the current strategy of declaring filters.
+This wouldn't require any change but supporting a new filtering language after the release of Haystack 2.x would be more difficult than doing it now.
```

Comment:
That's indeed an option that I considered, I'm not agaisn't it. 
I would make it clear that the old definition is deprecated though and push people toward the new definition.

Reply:
Just FYI: today thanks to @therve I discovered glom (https://glom.readthedocs.io/en/latest/tutorial.html), and I think it may help us implementing a translation layer between the old and the new syntax. I don't know if you're familiar with it, but at a quick glance it looks very interesting.